src/react/databrowser/objects/__tests__/Tags.test.jsx

```diff
@@ -0,0 +1,91 @@
+import * as s3objects from '../../../actions/s3object';
+import { AddButton, Item, SubButton } from '../../../ui-elements/EditableKeyValue';
+import { OBJECT_METADATA } from '../../../actions/__tests__/utils/testUtil';
+import React from 'react';
+import Tags from '../details/Tags';
+import { reduxMount } from '../../../utils/test';
+
+describe('Tags', () => {
+    const putObjectTaggingMock = jest.spyOn(s3objects, 'putObjectTagging').mockReturnValue({ type: '' });
+
+    afterEach(() => {
+        jest.clearAllMocks();
+    });
+
+    const tests = [
+        'Tags should render',
+        'should render by default an Item with empty values in each input when there are no key/value present',
+        'should add new key/value tag and should trigger function if save button is pressed',
+        'remove button and add button should be disabled',
+    ];
+
+    tests.forEach((t, index) => {
+        it(t, async () => {
+            const { component } = reduxMount(<Tags objectMetadata={OBJECT_METADATA}/>);
+
+            if (index === 0) {
+                expect(component.find(Tags).isEmptyRender()).toBe(false);
+            } else if (index === 1) {
+                expect(component.find('input#tags-input-key').prop('value')).toBe('');
+                expect(component.find('input#tags-input-value').prop('value')).toBe('');
+                expect(component.find(Item)).toHaveLength(1);
+            } else if (index === 2) {
+                // fill input key and input value
+                component.find('input#tags-input-key').simulate('change', { target: { name: 'key', value: 'key1' } });
+                component.find('input#tags-input-value').simulate('change', {
+                    target: {
+                        name: 'value',
+                        value: 'value1',
+                    },
+                });
+
+                // click on add button
+                component.find(AddButton).simulate('click');
+
+                // check new key/value added
+                expect(component.find(Item)).toHaveLength(2);
+
+                // check if function when pressing save button is triggered
+                expect(putObjectTaggingMock).toHaveBeenCalledTimes(0);
+                component.find('Button#tags-button-save').simulate('click');
+                expect(putObjectTaggingMock).toHaveBeenCalledTimes(1);
+            } else if (index === 3) {
+                expect(component.find(AddButton).children().prop('disabled')).toBe(true);
+                expect(component.find(SubButton).children().prop('disabled')).toBe(true);
+            }
+        });
+    });
+
+    it('should render an Item with key/value pass in props', () => {
+        const { component } = reduxMount(<Tags
+            objectMetadata={{ ...OBJECT_METADATA, tags: [{ key: 'key1', value: 'value1' }] }}/>);
+
+        expect(component.find('input#tags-input-key').prop('value')).toBe('key1');
```

Comment:
We can first check that only one "row" is returned before checking its value.

Reply:
done